File: python/lib/telegram_notifier.py
#!/usr/bin/env python3
"""
Telegram notification manager for the Issue Orchestrator.

This module handles sending notifications to Telegram about issue processing
status, errors, and completion.
"""

import logging
from telegram import Bot
from telegram.error import TelegramError


class TelegramNotifier:
    """Telegram notification manager"""
    
    def __init__(self, bot_token: str, chat_id: str):
        self.bot = Bot(token=bot_token)
        self.chat_id = chat_id
    
    async def send_message(self, message: str, parse_mode: str = "Markdown"):
        """Send a message to Telegram"""
        try:
            # Truncate message if too long
            max_length = 4000
            if len(message) > max_length:
                message = message[:max_length] + "\n... (truncated)"
            
            await self.bot.send_message(
                chat_id=self.chat_id,
                text=message,
                parse_mode=parse_mode
            )
        except TelegramError as e:
            logging.error(f"Failed to send Telegram message: {e}")
    
    async def notify_start(self, issue_number: int, title: str):
        """Notify that processing has started"""
        message = f"🚀 *Started issue #{issue_number}*: {title}"
        await self.send_message(message)
    
    async def notify_complete(self, issue_number: int, duration: str):
        """Notify that processing is complete"""
        message = f"✅ *Completed issue #{issue_number}* [{duration}]"
        await self.send_message(message)
    
    async def notify_needs_input(self, issue_number: int, output: str):
        """Notify that Claude needs input"""
        message = f"⏳ *Issue #{issue_number} needs input*:\n```\n{output[-1000:]}\n```"
        await self.send_message(message)
    
    async def notify_error(self, issue_number: int, error: str, output: str = None):
        """Notify about an error"""
        message = f"❌ *Error on issue #{issue_number}*:\n{error}"
        if output:
            message += f"\n\nLast output:\n```\n{output[-500:]}\n```"
        await self.send_message(message)

File: src/notifiers/telegramNotifier.ts
interface TelegramPayload {
  chat_id: string;
  text: string;
  parse_mode?: string;
}

export class TelegramNotifier {
  private readonly botToken: string;
  private readonly chatId: string;

  constructor(botToken: string, chatId: string) {
    this.botToken = botToken;
    this.chatId = chatId;
  }

  private async sendMessageBody(body: TelegramPayload): Promise<void> {
    const maxLength = 4000;
    if (body.text.length > maxLength) {
      body.text = `${body.text.slice(0, maxLength)}\n... (truncated)`;
    }

    const response = await fetch(`https://api.telegram.org/bot${this.botToken}/sendMessage`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      console.error(`Failed to send Telegram message: ${response.status}`);
    }
  }

  async sendMessage(message: string, parseMode: string = "Markdown"): Promise<void> {
    if (!this.botToken || !this.chatId) return;

    await this.sendMessageBody({
      chat_id: this.chatId,
      text: message,
      parse_mode: parseMode,
    });
  }

  async notifyStart(issueNumber: number, title: string): Promise<void> {
    await this.sendMessage(`🚀 *Started issue #${issueNumber}*: ${title}`);
  }

  async notifyComplete(issueNumber: number, duration: string): Promise<void> {
    await this.sendMessage(`✅ *Completed issue #${issueNumber}* [${duration}]`);
  }

  async notifyNeedsInput(issueNumber: number, output: string): Promise<void> {
    const snippet = output.slice(-1000);
    await this.sendMessage(`⏳ *Issue #${issueNumber} needs input*:\n\`\`\`\n${snippet}\n\`\`\``);
  }

  async notifyError(issueNumber: number, error: string, output?: string): Promise<void> {
    let message = `❌ *Error on issue #${issueNumber}*:\n${error}`;
    if (output) {
      const snippet = output.slice(-500);
      message += `\n\nLast output:\n\`\`\`\n${snippet}\n\`\`\``;
    }
    await this.sendMessage(message);
  }
}
